### PR TITLE
feat(HACBS-470): refactor code to add ReleaseStrategy params

### PR DESCRIPTION
**Pull request by davidmogar**
This change adds some changes into the pipeline_run.go file so it's possible to create empty ReleasePipelineRuns (without spec). To actually pass information to the PipelineRun, With* functions can be used. The new functions allow to specify the release labels, the pipeline reference, parameters, etc.

Signed-off-by: David Moreno García <damoreno@redhat.com>

### CODE REVIEW

File: config/crd/bases/appstudio.redhat.com_releasestrategies.yaml [hunk 1]
```diff
@@ -36,20 +36,39 @@ spec:
...
+                  description: Param declares an ArrayOrString to use for the parameter
+                    called name.
                   properties:
                     name:
                       type: string
                     value:
-                      items:
-                        type: string
-                      type: array
+                      description: ArrayOrString is a type that can hold a single
```
[the remark this answers is not in the document]
Comment (davidmogar) [reply config/crd/bases/appstudio.redhat.com_releasestrategies.yaml:51]: Similarly to the other ones, they come from the code documentation. In this case, those comments are in the upstream Tekton PipelineRun CRD.

File: tekton/pipeline_run.go [hunk 2]
```diff
@@ -48,24 +49,56 @@ var (
 	ReleaseWorkspaceLabel = fmt.Sprintf("%s/%s", releaseLabelPrefix, "workspace")
 )
 
-// CreateReleasePipelineRun creates a PipelineRun from a given Release and ReleaseStrategy.
-// Release details are added to the labels of the new PipelineRun to be able to reference it later on.
-func CreateReleasePipelineRun(release *v1alpha1.Release, releaseStrategy *v1alpha1.ReleaseStrategy) *tektonv1beta1.PipelineRun {
-	return &tektonv1beta1.PipelineRun{
+type ReleasePipelineRun struct {
```
Comment (johnbieren) [inline tekton/pipeline_run.go:52]: It is not clear to me why we are defining this struct instead of just using the normal `tektonv1beta1.PipelineRun`

File: controllers/release/release_adapter.go [hunk 3]
```diff
@@ -279,7 +282,7 @@ func (a *Adapter) getTargetReleaseLink() (*v1alpha1.ReleaseLink, error) {
 // registerReleasePipelineRunStatus updates the status of the Release being processed by monitoring the status of the
 // associated release PipelineRun and setting the appropriate state in the Release. If the PipelineRun hasn't
 // started/succeeded, no action will be taken.
-func (a *Adapter) registerReleasePipelineRunStatus(pipelineRun *v1beta1.PipelineRun) error {
+func (a *Adapter) registerReleasePipelineRunStatus(pipelineRun *tekton.ReleasePipelineRun) error {
```
Comment (johnbieren) [inline controllers/release/release_adapter.go:285]: See my comment pin `pipeline_run.go` about the new type but if we are using a new type, can't we just change the one last reference to `v1beta1` and remove the import entirely from this file?

File: tekton/pipeline_run.go (hunk 2 above)
Comment (davidmogar) [reply tekton/pipeline_run.go:52]: You cannot add new methods in a different file from where the receiver is defined.

File: controllers/release/release_adapter.go [hunk 4]
```diff
@@ -161,8 +161,11 @@ func (a *Adapter) EnsureReleasePipelineStatusIsTracked() (results.OperationResul
 
 // createReleasePipelineRun creates and returns a new release PipelineRun. The new PipelineRun will include owner
 // annotations, so it triggers Release reconciles whenever it changes.
```
Comment (johnbieren) [inline controllers/release/release_adapter.go:163]: I'd like a quick couple words about release strategy in the comment too

File: controllers/release/release_adapter.go (hunk 3 above)
Comment (davidmogar) [reply controllers/release/release_adapter.go:285]: I don’t really get this comment, sorry. Could you rephrase it?

File: config/crd/bases/appstudio.redhat.com_releasestrategies.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (johnbieren) [reply config/crd/bases/appstudio.redhat.com_releasestrategies.yaml:51]: Oh that's sweet

File: controllers/release/release_adapter.go (hunk 4 above)
Comment (davidmogar) [reply controllers/release/release_adapter.go:163]: Will change it.

File: tekton/pipeline_run.go (hunk 2 above)
Comment (johnbieren) [reply tekton/pipeline_run.go:52]: That's unfortunate. Can you comment pretty much exactly that before defining this then?

Comment (davidmogar) [reply tekton/pipeline_run.go:52]: Can do.

File: controllers/release/release_adapter.go (hunk 3 above)
Comment (davidmogar) [reply controllers/release/release_adapter.go:285]: Oh, got you now. I was confused because the discussion was for a different line.

No, I don't think it's possible to do that as a ReleasePipelineRun is not a PipelineRun (it's something more), so listing would not work.

Comment (johnbieren) [reply controllers/release/release_adapter.go:285]: I thought the same but figured it was worth asking the go expert

Comment (davidmogar) [reply controllers/release/release_adapter.go:285]: I’d like to meet that guy too.

File: tekton/pipeline_run.go [hunk 5]
```diff
@@ -48,24 +49,69 @@ var (
...
+
+	return &ReleasePipelineRun{pipelineRun}
+}
+
+// AsPipelineRun casts the ReleasePipelineRun to PipelineRun, so it can be used in the Kubernetes client.
+func (r *ReleasePipelineRun) AsPipelineRun() *tektonv1beta1.PipelineRun {
+	return &r.PipelineRun
+}
+
+// WithExtraParam adds an extra param to the release PipelineRun.
```
Comment (johnbieren) [inline tekton/pipeline_run.go:75]: What do you think about adding a comment along the lines of "If the parameter is not part of the invoked release Pipeline definition, it is silently ignored"

I say this just because when I was implementing this, it is something I thought of and tested

**Review by johnbieren — APPROVED** [review]
One small comment but otherwise lgtm
